### PR TITLE
Test current code base as an integration test for PRs and pushes

### PR DESCRIPTION
**Pull request by Pike**
**Description**
> Add a build step to create lib and node_modules artifact
> Add an integration test to run on PRs

**Testing Instructions**
> Happens as part of this PR.

**Additional Notes**
> I bet the integration test will fail for insufficient permissions. We'll see.

### CODE REVIEW

**Comment by Pike** [comment]
So, this failed in a good way.

Things I like about this PR: It indicates that just installing the prod dependencies should do the trick for the action (3.7.1 ships dev dependencies, too). It runs an integration test.

I'm tempted to stack `dry-run` on top to allow for the integration test to actually pass :-) . @JamesIves ?

**Comment by JamesIves** [comment]
I think adding the dry run stuff to this PR will be the only way this will work unfortunately, but I'm a big fan of this. It should really help! I'm thinking we should expand this to run the entire integration test suite and not just a single one.

The release branch gets updated manually by me whenever I merge something into it (I just switch and run `yarn install`). I've been wanting to automate that process as the action could in theory deploy the dependencies its self in a similar manor to this whenever content is merged into that branch.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -34,3 +34,60 @@ jobs:
         uses: codecov/codecov-action@v1
         with:
           token: ${{ secrets.CODECOV_TOKEN }}
+
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v1
```
Comment (JamesIves) [inline .github/workflows/build.yml:42]: We should run `v2` here with `persist-credentials: false` as it's the latest version; https://github.com/JamesIves/github-pages-deploy-action/blob/dev/.github/workflows/integration.yml#L42-L45

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -34,3 +34,60 @@ jobs:
...
+        uses: actions/download-artifact@v2
+        with:
+          name: dist
+
+      - name: Deploy
+        uses: ./
+        with:
+          FOLDER: integration
+          BASE_BRANCH: ${{ github.sha }}
+          BRANCH: gh-pages-pr
```
Comment (JamesIves) [inline .github/workflows/build.yml:92]: We should consider adding a cleanup step here in the off chance that it does actually generate these branches (ie if someone with adequate permissions made a PR). This is a potential window for abuse so it's probably best we prevent things from lingering: https://github.com/JamesIves/github-pages-deploy-action/blob/dev/.github/workflows/integration.yml#L31-L35

**Comment by Pike** [comment]
I've spun dry-run into its own PR #526 , with some unit tests.

With that, I wonder if I should move the steps into the `integration.yml` workflow completely?

**Comment by JamesIves** [comment]
Lemme gather my thoughts on this, I'll take a closer look later today!

**Comment by JamesIves** [comment]
Quick question before I merge: Should this include the dry-run flag? I think keeping this in build/PR makes sense as the intention of this is to smoke test PR's instead of running the scheduled tests.

**Comment by Pike** [comment]
Soooo, `--dry-run` isn't as dry run as I thought it was, it does fail for failed permissions. I filed #536 on that, should I tackle this as part of this PR? Sounds like doing it here is the best to avoid more chickens and eggs?

**Comment by JamesIves** [comment]
Yeah tackling it as part of this PR makes the most sense to me!

**Comment by Pike** [comment]
Nice, I'm finding actual bugs now. Also bugs I introduced.

**Comment by Pike** [comment]
For reference, `generateBranch` runs in `workspace`, but needs to run in `temporaryDeploymentDirectory`, now that we don't mess with `workspace` anymore. More refactorings coming up, sorry for that.

**Comment by Pike** [comment]
I'm putting this into draft state for now, I need to do a couple of iterations here, probably.

I'll update the PR here soon, but I'm also changing a lot, so feel free to comment and complain liberally.

The high-level idea is that the mix of `worktree` and `generateBranch` and `singleCommit` turned out to be hard to untangle for me. So I'm switching over to a `generateWorktree`, and then a rather different logic for `singleCommit`. Basically, I'm starting off with an orphan branch right away, and then use `hg diff` to figure out if to actually commit or not.

**Comment by JamesIves** [comment]
Sounds good to me, I'll keep an eye out. I've got a lot of spare time coming up this weekend so I'm hoping to knock out a lot of these remaining tasks for v4.

**Comment by Pike** [comment]
I think this is in a shape where I'd appreciate feedback. Candid-european is perfectly fine.

A few thoughts I have:

I think the dry-run integration testing turned out rather well. I wish it was more strict, which might be be OK to do with checking the action output. Right now it's an environment, in particular with multiple deploy steps, that's interesting. Is there a reason to not make it an step output? I also entertained the idea to expose the count of commits done as step output, which could be checked for things like `singleCommit` or not.

I introduced an abstraction of a git command. It's not placed anywhere logical, but just close to where it's used. I like being able to set options one by one, and then have the class generate the command string. I do think I only did this half-way, though. `execute` could probably take  that abstraction|string as argument, and `toString()` itself. I'm not sure if this pays off in other parts of the code. In `generateWorktree`, there's a mix between `--orphan` and `-B` that might be more confusing to write as a string composition..

I also like the actual testing of `createWorktree` with actual git content. With that being properly (IMHO) tested by itself, it'd be nice to mock it in `git.test.ts`, but I haven't found a way to do that.

Also, I'm a bit lost on how to test the error handling in `createWorktree`.

**Comment by JamesIves** [comment]
I will take a look first thing tomorrow.

**Review by JamesIves — COMMENTED** [review]
This looks really good, and I really appreciate you taking the time on this. I have a few comments/questions.

Regarding testing the error handling there's a way you can make the `execute` function throw if it's given a specific string, I believe it's done here: https://github.com/JamesIves/github-pages-deploy-action/blob/dev/__tests__/git.test.ts#L56-L58 

As we can't necessarily test git failures just ensuring that we have coverage on the catch would be significant, that way we know it will bubble up to the calling function and provide the user with an adequate error message.

File: __tests__/worktree.test.ts [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    tempdir = fs.mkdtempSync(path.join(os.tmpdir(), 'gh-deploy-'))
+    const origin = path.join(tempdir, 'origin')
+    await execute('git init origin', tempdir, silent)
+    await execute('git config user.email "you@example.com"', origin, silent)
+    await execute('git config user.name "Jane Doe"', origin, silent)
+    await execute('git checkout -b main', origin, silent)
+    fs.writeFileSync(path.join(origin, 'f1'), 'hello world\n')
+    await execute('git add .', origin, silent)
+    await execute('git commit -mc0', origin, silent)
+    fs.writeFileSync(path.join(origin, 'f1'), 'hello world\nonce more\n')
```
[the remark this answers is not in the document]
Comment (Pike) [reply __tests__/worktree.test.ts:29]: See, this happens when you're not a native speaker. Not really a typo, I'll just use  `\nand planets` instead.

File: __tests__/worktree.test.ts [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    await execute('git reset --hard', origin, silent)
+    await fs.promises.writeFile(path.join(origin, 'gh1'), 'pages content\n')
+    await execute('git add .', origin, silent)
+    await execute('git commit -mgh0', origin, silent)
+    await fs.promises.writeFile(
+      path.join(origin, 'gh1'),
+      'pages content\ngoes on\n'
+    )
+    await execute('git add .', origin, silent)
+    await execute('git commit -mgh1', origin, silent)
```
[the remark this answers is not in the document]
Comment (Pike) [reply __tests__/worktree.test.ts:42]: Yeah, I'm adding a bunch of comments to the test file now.

**Comment by Pike** [comment]
I'm going to just add an explicit test file for error handing. The problem (AFAICT) is that you can only mock modules per test file. So I'm having one test file that doesn't mock `execute.ts`, and one test file that does.

**Comment by Pike** [comment]
Thanks for the review, I think I've got them all.

I'm not all that fond of https://github.com/JamesIves/github-pages-deploy-action/pull/505/commits/c3cddcb259310b18db97e9f2cb6697c596e19d4e, but I also don't have a better idea. W/out that, coverage complains about the conditional check for changes files. I made that a separate commit so it's easier to mess with.

**Comment by Pike** [comment]
Another idea I just had was to mock

```
jest.mock('../src/worktree', () => ({
  // eslint-disable-next-line @typescript-eslint/naming-convention
  __esModule: true,
  generateWorktree: jest.fn()
}))
```

in `git.test.ts`. That's an opportunity now that `generateWorktree` is in its  own module.That would decouple the tests further, in particular if the commands in `generateWorktree` changed. Not sure it does much more than that. Well, it cuts down the different call counts to `execute` from 8, 10, 11, 12 to 4, 7, 8

File: __tests__/worktree.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (JamesIves) [reply __tests__/worktree.test.ts]: haha yeah totally understand

**Comment by JamesIves** [comment]
> Thanks for the review, I think I've got them all.
> 
> I'm not all that fond of [c3cddcb](https://github.com/JamesIves/github-pages-deploy-action/commit/c3cddcb259310b18db97e9f2cb6697c596e19d4e), but I also don't have a better idea. W/out that, coverage complains about the conditional check for changes files. I made that a separate commit so it's easier to mess with.

There is already a `isTest` variable, would that be appropriate instead of adding another variable?

**Comment by Pike** [comment]
> There is already a `isTest` variable, would that be appropriate instead of adding another?

I've been wondering about that a bit. Right now, `isTest` is basically `hasChangedFilesForTest`?

I'm not too fond of using the same value for both test flags, but we could use an enum like

```
enum TestFlag {
  NONE = 0,
  HAS_CHANGED_FILES = 1 << 1,
  HAS_REMOTE_BRANCH = 1 << 2,
}
```

Than we only need one variable, but can combine different scenarios in one test. It's also explicit when setting up the test?

**Comment by JamesIves** [comment]
I like the use of the enum!

**Comment by Pike** [comment]
I have mixed feelings on whether a required or an optional `isTest` flag is worse. TS complains about `action.isTest & TestFlag.HAS_CHANGED_FILES`  if it's optional, which makes the code harder to read if we'd work around that. And requiring it makes the tests more verbose.

The `.eslintrc.json` change is something I found on the net when trying to fix the new exception,  and it seems to work w/out exceptions and details now.

**Comment by JamesIves** [comment]
I see what you mean, it is somewhat confusing to have the two. I think for now let's merge this, and in a follow-up figure out if we can deprecate the use case of `isTest` as it's a bit of a cheeky workaround to begin with anyway.

**Comment by JamesIves** [comment]
Now in beta 2: https://github.com/JamesIves/github-pages-deploy-action/releases/tag/4.0.0-beta-02

**Pull request merged**